libcudacxx/include/cuda/std/detail/libcxx/include/__complex/nvbf16.h

```diff
@@ -0,0 +1,131 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _LIBCUDACXX___COMPLEX_NVBF16_H
+#define _LIBCUDACXX___COMPLEX_NVBF16_H
+
+#ifndef __cuda_std__
+#  include <config>
+#endif // __cuda_std__
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#if defined(_LIBCUDACXX_HAS_NVBF16)
+
+_CCCL_DIAG_PUSH
+_CCCL_DIAG_SUPPRESS_CLANG("-Wunused-function")
+#  include <cuda_bf16.h>
+_CCCL_DIAG_POP
+
+#  include "../__complex/traits.h"
+#  include "../__cuda/cmath_nvbf16.h"
+#  include "../__type_traits/integral_constant.h"
+#  include "../cmath"
+#  include "../complex"
+
+#  if !defined(_CCCL_COMPILER_NVRTC)
+#    include <sstream> // for std::basic_ostringstream
+#  endif // !_CCCL_COMPILER_NVRTC
+
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+
+template <>
+struct __is_nvbf16<__nv_bfloat16> : true_type
+{};
+
+template <>
+struct __complex_alignment<__nv_bfloat16> : integral_constant<size_t, alignof(__nv_bfloat162)>
+{};
+
+template <>
+struct __type_to_vector<__nv_bfloat16>
+{
+  using __type = __nv_bfloat162;
+};
+
+template <>
+struct __libcpp_complex_overload_traits<__nv_bfloat16, false, false>
+{
+  typedef __nv_bfloat16 _ValueType;
+  typedef complex<__nv_bfloat16> _ComplexType;
+};
+
+// We can utilize vectorized operations for those operators
+inline _LIBCUDACXX_INLINE_VISIBILITY complex<__nv_bfloat16>&
+operator+=(complex<__nv_bfloat16>& __lhs, const complex<__nv_bfloat16>& __rhs) noexcept
+{
+  reinterpret_cast<__nv_bfloat162&>(__lhs) += reinterpret_cast<const __nv_bfloat162&>(__rhs);
```

Comment:
Okay, this is definitely sus. I mean _maybe_ we can depend on the compiler not being malicious enough and doing initial common subsequence analysis despite complex not fulfilling the requirements for it to kick in, and despite this not being a union? This just feels like a worse violation of strict aliasing than reinterpreting the two members as an array.

I'd be much more comfortable if we actually stored the vector type in vector-enabled cases of complex, though I realize that that'd make us no longer able to simply access `__re_`...

I wonder if the codegen was really bad if, instead of breaking the aliasing here, we had an assignment from the vector type, and did `__lhs = __nvbfloat162{__lhs.__re_, __lhs.__im_} + __nvbfloat162{__rhs.__re_, __rhs.__im_};`...

Reply:
Actually, I think there's a way to make this _roughly_ legit, and it only involves modifying the storage definition. For non-vectorable types, keep just the two member defs.

When a vector type is available though, we define storage as:

```c++
union {
  struct {
    _Tp __re_;
    _Tp __im_;
  };
  vector_type<_Tp> __vect_;
};
```

Then we can even define the += operator and friends in the primary template, dispatching to code that either adds `__vect_`s together, or that does this piecewise, depending on whether there's a vector representation; moreover, at `constexpr` it only uses `__re_` and `__im_`, because accessing the inactive member of a union is not allowed at compile time. But because of the common initial subsequence rules, I _think_ the runtime accesses to `__vect_` are _mostly_ fine even if we initialized the non-vector part of the representation?

I know that the above kinda _looks_ wrong, but (1) I'm reasonably certain it fixes my suspicions about this code, and (2) allows us to generically define the vector cases near the main template, so that we don't have to redefine them over and over again when we get and/or add support for more types with vector representations available.

Reply:
That is what I had in between, however it forces us to essentially replicate everything for the vector types, as we need to reimplement each and every method.

The original issue I had is from trying to make it an aggregate, which is super awkward with a vector type as one cannot write `complex meow{1.0, 2.0}` but needs to write `complex meow{{1.0, 2.0}}`

I will explore using a separate definition. Maybe we can trick the compiler by making all the operators hidden friends?
Another option would be to use a bitcast when possible? We could use the temporary for old compilers and bitcast for newer ones

Reply:
So bit_cast wont work because our vector types are not trivially_copyable and the union does not work because of the same reason

Reply:
So I did something, that might even work:
```cpp
union __complex_bfloat_vector_op
{
  struct
  {
    complex<__nv_bfloat16>& __as_complex_lhs;
    const complex<__nv_bfloat16>& __as_complex_rhs;
  };
  struct
  {
    __nv_bfloat162& __as_vector_lhs;
    const __nv_bfloat162& __as_vector_rhs;
  };
};
```

This lets us define a wrapper around references to those objects and switches between those. Its morally the same but no reinterpretation at least.

We can then do the various operations as:
```cpp
  _LIBCUDACXX_INLINE_VISIBILITY complex<__nv_bfloat16>& __plus_op() && noexcept
  {
    __as_vector_lhs += __as_vector_rhs;
    return __as_complex_lhs;
  }
```